### PR TITLE
Report handler features for KEP-127 and KEP-3857

### PR DESCRIPTION
**Pull request by giuseppe**
needs the changes in: https://github.com/kubernetes/kubernetes/pull/123216

#### What type of PR is this?

/kind feature

#### Does this PR introduce a user-facing change?

```release-note
Now CRI-O reports to the Kubelet what features are available to a handler
```

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -1595,6 +1595,19 @@ func (r *RuntimeHandler) RuntimeSupportsIDMap() bool {
 	return true
 }
 
+// RuntimeSupportsMountFlag returns whether this runtime supports the specified mount option.
+func (r *RuntimeHandler) RuntimeSupportsMountFlag(flag string) bool {
+	if r.features.MountOptions == nil {
+		return false
+	}
+	for _, m := range r.features.MountOptions {
```
Comment (rata) [inline pkg/config/config.go:1603]: You can use the "slices" package and remove the loop. It seems crio is using go 1.21 already, that is the first release to include `slices.Contains()`

Comment (giuseppe) [reply pkg/config/config.go]: good hint, fixed now

File: go.mod [hunk 2]
```diff
@@ -75,7 +75,7 @@ require (
 	k8s.io/api v0.29.1
 	k8s.io/apimachinery v0.29.1
 	k8s.io/client-go v0.29.1
-	k8s.io/cri-api v0.29.1
+	k8s.io/cri-api v0.30.0-alpha.2.0.20240217224521-840a52e4cd66
 	k8s.io/klog/v2 v2.120.1
 	k8s.io/kubelet v0.29.1
```
Comment (saschagrunert) [inline go.mod:80]: Can we bump all of them to [v1.30.0-alpha.2](https://github.com/kubernetes/kubernetes/tree/v1.30.0-alpha.2)?

**Comment by giuseppe** [comment]
@saschagrunert are these CI failures expected?

**Review by mrunalp — APPROVED** [review]
/lgtm

File: server/runtime_status.go [hunk 3]
```diff
@@ -36,6 +36,30 @@ func (s *Server) Status(ctx context.Context, req *types.StatusRequest) (*types.S
...
+				},
+			}
+		}
+
+		rro := runtime.RuntimeSupportsMountFlag("rro")
+		userns := runtime.RuntimeSupportsIDMap()
+		h := makeRuntimeHandler(name, rro, userns)
+		resp.RuntimeHandlers = append(resp.RuntimeHandlers, h)
+
+		// if it is the default runtime, also add it with an empty name
```
Comment (kwilczynski) [inline server/runtime_status.go:56]:
```suggestion
		// If it's the default runtime, add it with an empty name.
```

File: server/runtime_status.go [hunk 4]
```diff
@@ -36,6 +36,30 @@ func (s *Server) Status(ctx context.Context, req *types.StatusRequest) (*types.S
...
+			return &types.RuntimeHandler{
+				Name: name,
+				Features: &types.RuntimeHandlerFeatures{
+					RecursiveReadOnlyMounts: rro,
+					UserNamespaces:          userns,
+				},
+			}
+		}
+
+		rro := runtime.RuntimeSupportsMountFlag("rro")
```
Comment (AkihiroSuda) [inline server/runtime_status.go:51]: This shouldn't be set until CRI-O actually implements the feature

Comment (AkihiroSuda) [reply server/runtime_status.go:51]: critest (https://github.com/kubernetes-sigs/cri-tools/pull/1344) is failing due to this

Comment (kwilczynski) [reply server/runtime_status.go:51]: Temporary workaround using:

- https://github.com/cri-o/cri-o/pull/7828

Fixed via:

- https://github.com/cri-o/cri-o/pull/7962

**Pull request merged**